File: PhotographerApp.Infrastructure/Services/Business/FileStorageService.cs
using SixLabors.ImageSharp;
using SixLabors.ImageSharp.Processing;
using PhotographerApp.Core.Interfaces;
using Microsoft.AspNetCore.Http;

namespace PhotographerApp.Infrastructure.Services.Business;

public class FileStorageService : IFileStorageService
{
    private readonly string _basePath;
    private const int MaxImageSizeBytes = 10 * 1024 * 1024;
    private const int MaxVideoSizeBytes = 200 * 1024 * 1024;
    private const int ThumbnailWidth = 400;
    // Allow all common image extensions
    private static readonly string[] AllowedImageExtensions =
    {
        ".jpg", ".jpeg", ".png", ".gif", ".bmp", ".tiff", ".tif", ".webp", ".ico", ".svg", ".heic", ".heif"
    };
    private static readonly string[] AllowedVideoExtensions = { ".mp4", ".webm", ".mov", ".avi", ".mkv", ".flv", ".wmv" };
    private static readonly byte[][] ImageMagicNumbers = new[]
    {
        new byte[] { 0xFF, 0xD8, 0xFF },                           // JPEG
        new byte[] { 0x89, 0x50, 0x4E, 0x47 },                     // PNG
        new byte[] { 0x47, 0x49, 0x46, 0x38 },                     // GIF
        new byte[] { 0x42, 0x4D },                                 // BMP
        new byte[] { 0x49, 0x49, 0x2A, 0x00 },                     // TIFF (little-endian)
        new byte[] { 0x4D, 0x4D, 0x00, 0x2A },                     // TIFF (big-endian)
        new byte[] { 0x52, 0x49, 0x46, 0x46 },                     // WEBP
        new byte[] { 0x00, 0x00, 0x01, 0x00 },                     // ICO
        new byte[] { 0x00, 0x00, 0x00, 0x18, 0x66, 0x74, 0x79, 0x70 } // HEIC/HEIF
    };
    private static readonly byte[][] VideoMagicNumbers = new[]
    {
        new byte[] { 0x00, 0x00, 0x00, 0x20, 0x66, 0x74, 0x79, 0x70 }, // MP4
        new byte[] { 0x1A, 0x45, 0xDF, 0xA3 },                          // MKV
        new byte[] { 0x52, 0x49, 0x46, 0x46 },                          // AVI/WEBM
        new byte[] { 0x00, 0x00, 0x00, 0x14, 0x66, 0x74, 0x79, 0x70 }   // MOV
    };

    public FileStorageService(string webRootPath = null)
    {
        var uploadsPath = webRootPath ?? Path.Combine(Directory.GetCurrentDirectory(), "wwwroot");
        _basePath = Path.Combine(uploadsPath, "uploads");
        EnsureDirectoriesExist();
    }

    public async Task<string> SavePhotoAsync(Stream fileStream, string originalFileName, CancellationToken cancellationToken = default)
    {
        if (fileStream == null)
            throw new ArgumentNullException(nameof(fileStream));
        if (string.IsNullOrWhiteSpace(originalFileName))
            throw new ArgumentException("File name cannot be empty", nameof(originalFileName));

        var fileName = GenerateFileName(originalFileName);
        var yearMonth = $"{DateTime.UtcNow:yyyy/MM}";
        var photoDir = Path.Combine(_basePath, "photos", yearMonth);
        Directory.CreateDirectory(photoDir);

        var filePath = Path.Combine(photoDir, fileName);
        using (var fileToWrite = File.Create(filePath))
        {
            await fileStream.CopyToAsync(fileToWrite, cancellationToken);
        }

        return $"/uploads/photos/{yearMonth}/{fileName}";
    }

    public async Task<string> SavePhotoThumbnailAsync(Stream fileStream, string originalFileName, CancellationToken cancellationToken = default)
    {
        if (fileStream == null)
            throw new ArgumentNullException(nameof(fileStream));
        if (string.IsNullOrWhiteSpace(originalFileName))
            throw new ArgumentException("File name cannot be empty", nameof(originalFileName));

        var fileName = GenerateFileName(originalFileName);
        var yearMonth = $"{DateTime.UtcNow:yyyy/MM}";
        var photoDir = Path.Combine(_basePath, "photos", yearMonth);
        Directory.CreateDirectory(photoDir);

        var thumbnailFileName = Path.GetFileNameWithoutExtension(fileName) + "_thumb" + Path.GetExtension(fileName);
        var thumbnailPath = Path.Combine(photoDir, thumbnailFileName);

        fileStream.Seek(0, SeekOrigin.Begin);

        using (var image = await Image.LoadAsync(fileStream, cancellationToken))
        {
            var aspectRatio = (double)image.Width / image.Height;
            var newHeight = (int)(ThumbnailWidth / aspectRatio);

            image.Mutate(x => x.Resize(ThumbnailWidth, newHeight, KnownResamplers.Lanczos3));

            await image.SaveAsJpegAsync(thumbnailPath, cancellationToken: cancellationToken);
        }

        return $"/uploads/photos/{yearMonth}/{thumbnailFileName}";
    }

    public async Task<(int Width, int Height)> GetImageDimensionsAsync(Stream fileStream, CancellationToken cancellationToken = default)
    {
        if (fileStream == null)
            throw new ArgumentNullException(nameof(fileStream));

        fileStream.Seek(0, SeekOrigin.Begin);
        using (var image = await Image.LoadAsync(fileStream, cancellationToken))
        {
            return (image.Width, image.Height);
        }
    }

    public async Task<string> SaveVideoAsync(Stream fileStream, string originalFileName, CancellationToken cancellationToken = default)
    {
        if (fileStream == null)
            throw new ArgumentNullException(nameof(fileStream));
        if (string.IsNullOrWhiteSpace(originalFileName))
            throw new ArgumentException("File name cannot be empty", nameof(originalFileName));

        var fileName = GenerateFileName(originalFileName);
        var yearMonth = $"{DateTime.UtcNow:yyyy/MM}";
        var videoDir = Path.Combine(_basePath, "videos", yearMonth);
        Directory.CreateDirectory(videoDir);

        var filePath = Path.Combine(videoDir, fileName);
        using (var fileToWrite = File.Create(filePath))
        {
            await fileStream.CopyToAsync(fileToWrite, cancellationToken);
        }

        return $"/uploads/videos/{yearMonth}/{fileName}";
    }

    public async Task<string> SaveVideoThumbnailAsync(Stream fileStream, string originalFileName, CancellationToken cancellationToken = default)
    {
        if (fileStream == null)
            throw new ArgumentNullException(nameof(fileStream));
        if (string.IsNullOrWhiteSpace(originalFileName))
            throw new ArgumentException("File name cannot be empty", nameof(originalFileName));

        var fileName = GenerateFileName(originalFileName);
        var yearMonth = $"{DateTime.UtcNow:yyyy/MM}";
        var videoDir = Path.Combine(_basePath, "videos", yearMonth);
        Directory.CreateDirectory(videoDir);

        var thumbnailFileName = Path.GetFileNameWithoutExtension(fileName) + "_thumb.jpg";
        var thumbnailPath = Path.Combine(videoDir, thumbnailFileName);

        fileStream.Seek(0, SeekOrigin.Begin);

        try
        {
            using (var image = await Image.LoadAsync(fileStream, cancellationToken))
            {
                var aspectRatio = (double)image.Width / image.Height;
                var newHeight = (int)(ThumbnailWidth / aspectRatio);

                image.Mutate(x => x.Resize(ThumbnailWidth, newHeight, KnownResamplers.Lanczos3));
                await image.SaveAsJpegAsync(thumbnailPath, cancellationToken: cancellationToken);
            }
        }
        catch (Exception ex)
        {
            // Return empty string if thumbnail generation fails
            System.Diagnostics.Debug.WriteLine($"Thumbnail generation failed: {ex.Message}");
            return string.Empty;
        }

        return $"/uploads/videos/{yearMonth}/{thumbnailFileName}";
    }

    public async Task DeleteFileAsync(string filePath, CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrEmpty(filePath))
            return;

        var fullPath = Path.Combine(Directory.GetCurrentDirectory(), "wwwroot", filePath.TrimStart('/'));
        if (File.Exists(fullPath))
        {
            File.Delete(fullPath);
        }

        await Task.CompletedTask;
    }

    public bool ValidateImageFile(IFormFile file, out string errorMessage)
    {
        errorMessage = string.Empty;

        if (file == null || file.Length == 0)
        {
            errorMessage = "File is empty";
            return false;
        }

        if (file.Length > MaxImageSizeBytes)
        {
            errorMessage = $"File size exceeds {MaxImageSizeBytes / (1024 * 1024)}MB limit";
            return false;
        }

        var extension = Path.GetExtension(file.FileName).ToLowerInvariant();
        if (!AllowedImageExtensions.Contains(extension))
        {
            errorMessage = "File type not allowed. Allowed types: jpg, jpeg, png, gif, bmp, tiff, webp, ico, svg, heic, heif";
            return false;
        }

        if (!ValidateMagicBytes(file, ImageMagicNumbers))
        {
            errorMessage = "File content does not match image format";
            return false;
        }

        return true;
    }

    public bool ValidateVideoFile(IFormFile file, out string errorMessage)
    {
        errorMessage = string.Empty;

        if (file == null || file.Length == 0)
        {
            errorMessage = "File is empty";
            return false;
        }

        if (file.Length > MaxVideoSizeBytes)
        {
            errorMessage = $"File size exceeds {MaxVideoSizeBytes / (1024 * 1024)}MB limit";
            return false;
        }

        var extension = Path.GetExtension(file.FileName).ToLowerInvariant();
        if (!AllowedVideoExtensions.Contains(extension))
        {
            errorMessage = "File type not allowed. Allowed types: mp4, webm, mov, avi, mkv, flv, wmv";
            return false;
        }

        if (!ValidateMagicBytes(file, VideoMagicNumbers))
        {
            errorMessage = "File content does not match video format";
            return false;
        }

        return true;
    }

    private bool ValidateMagicBytes(IFormFile file, byte[][] allowedMagicNumbers)
    {
        var buffer = new byte[8];
        file.OpenReadStream().Read(buffer, 0, 8);

        foreach (var magicNumber in allowedMagicNumbers)
        {
            if (buffer.Take(magicNumber.Length).SequenceEqual(magicNumber))
                return true;
        }

        return false;
    }

    private string GenerateFileName(string originalFileName)
    {
        var extension = Path.GetExtension(originalFileName);
        return $"{Guid.NewGuid().ToString("N")}{extension}";
    }

    private void EnsureDirectoriesExist()
    {
        Directory.CreateDirectory(Path.Combine(_basePath, "photos"));
        Directory.CreateDirectory(Path.Combine(_basePath, "videos"));
        Directory.CreateDirectory(Path.Combine(_basePath, "team"));
    }
}

File: src/files/file-storage.service.ts
import { randomUUID } from 'node:crypto';
import * as fs from 'node:fs/promises';
import * as path from 'node:path';
import { Injectable, OnModuleInit } from '@nestjs/common';
import sharp from 'sharp';

const MAX_IMAGE_SIZE_BYTES = 10 * 1024 * 1024;
const MAX_VIDEO_SIZE_BYTES = 200 * 1024 * 1024;
const THUMBNAIL_WIDTH = 400;

const ALLOWED_IMAGE_EXTENSIONS = ['.jpg', '.jpeg', '.png', '.gif', '.bmp', '.tiff', '.tif', '.webp', '.ico', '.svg', '.heic', '.heif'];
const ALLOWED_VIDEO_EXTENSIONS = ['.mp4', '.webm', '.mov', '.avi', '.mkv', '.flv', '.wmv'];

const IMAGE_MAGIC_NUMBERS: number[][] = [
  [0xff, 0xd8, 0xff], // JPEG
  [0x89, 0x50, 0x4e, 0x47], // PNG
  [0x47, 0x49, 0x46, 0x38], // GIF
  [0x42, 0x4d], // BMP
  [0x49, 0x49, 0x2a, 0x00], // TIFF (little-endian)
  [0x4d, 0x4d, 0x00, 0x2a], // TIFF (big-endian)
  [0x52, 0x49, 0x46, 0x46], // WEBP
  [0x00, 0x00, 0x01, 0x00], // ICO
  [0x00, 0x00, 0x00, 0x18, 0x66, 0x74, 0x79, 0x70], // HEIC/HEIF
];
const VIDEO_MAGIC_NUMBERS: number[][] = [
  [0x00, 0x00, 0x00, 0x20, 0x66, 0x74, 0x79, 0x70], // MP4
  [0x1a, 0x45, 0xdf, 0xa3], // MKV
  [0x52, 0x49, 0x46, 0x46], // AVI/WEBM
  [0x00, 0x00, 0x00, 0x14, 0x66, 0x74, 0x79, 0x70], // MOV
];

export interface UploadedFileLike {
  buffer: Buffer;
  originalname: string;
  size: number;
}

@Injectable()
export class FileStorageService implements OnModuleInit {
  private readonly basePath: string;

  constructor() {
    this.basePath = path.join(process.cwd(), 'uploads');
  }

  async onModuleInit(): Promise<void> {
    await Promise.all(
      ['photos', 'videos', 'team'].map((dir) => fs.mkdir(path.join(this.basePath, dir), { recursive: true })),
    );
  }

  async savePhoto(file: UploadedFileLike): Promise<string> {
    const { fileName, dir, yearMonth } = await this.preparePath(file.originalname, 'photos');
    await fs.writeFile(path.join(dir, fileName), file.buffer);
    return `/uploads/photos/${yearMonth}/${fileName}`;
  }

  async savePhotoThumbnail(file: UploadedFileLike): Promise<string> {
    const { fileName, dir, yearMonth } = await this.preparePath(file.originalname, 'photos');
    const thumbnailFileName = `${path.parse(fileName).name}_thumb.jpg`;

    const image = sharp(file.buffer);
    const metadata = await image.metadata();
    const aspectRatio = (metadata.width ?? THUMBNAIL_WIDTH) / (metadata.height ?? THUMBNAIL_WIDTH);
    const newHeight = Math.round(THUMBNAIL_WIDTH / aspectRatio);

    await image.resize(THUMBNAIL_WIDTH, newHeight).jpeg().toFile(path.join(dir, thumbnailFileName));

    return `/uploads/photos/${yearMonth}/${thumbnailFileName}`;
  }

  async getImageDimensions(buffer: Buffer): Promise<{ width: number; height: number }> {
    const metadata = await sharp(buffer).metadata();
    return { width: metadata.width ?? 0, height: metadata.height ?? 0 };
  }

  async saveVideo(file: UploadedFileLike): Promise<string> {
    const { fileName, dir, yearMonth } = await this.preparePath(file.originalname, 'videos');
    await fs.writeFile(path.join(dir, fileName), file.buffer);
    return `/uploads/videos/${yearMonth}/${fileName}`;
  }

  /** Mirrors FileStorageService.cs's SaveVideoThumbnailAsync: it tries to decode the video bytes as
   * a still image (ImageSharp can't handle video codecs), which fails for real videos and returns
   * "" — a pre-existing no-op, not something this migration is meant to fix. */
  async saveVideoThumbnail(file: UploadedFileLike): Promise<string> {
    try {
      const { fileName, dir, yearMonth } = await this.preparePath(file.originalname, 'videos');
      const thumbnailFileName = `${path.parse(fileName).name}_thumb.jpg`;

      const image = sharp(file.buffer);
      const metadata = await image.metadata();
      const aspectRatio = (metadata.width ?? THUMBNAIL_WIDTH) / (metadata.height ?? THUMBNAIL_WIDTH);
      const newHeight = Math.round(THUMBNAIL_WIDTH / aspectRatio);

      await image.resize(THUMBNAIL_WIDTH, newHeight).jpeg().toFile(path.join(dir, thumbnailFileName));
      return `/uploads/videos/${yearMonth}/${thumbnailFileName}`;
    } catch {
      return '';
    }
  }

  async deleteFile(filePath: string | null | undefined): Promise<void> {
    if (!filePath) return;
    const fullPath = path.join(this.basePath, filePath.replace(/^\/?uploads\//, ''));
    await fs.rm(fullPath, { force: true });
  }

  validateImageFile(file: UploadedFileLike): { valid: boolean; error?: string } {
    if (!file || file.size === 0) return { valid: false, error: 'File is empty' };
    if (file.size > MAX_IMAGE_SIZE_BYTES) {
      return { valid: false, error: `File size exceeds ${MAX_IMAGE_SIZE_BYTES / (1024 * 1024)}MB limit` };
    }

    const extension = path.extname(file.originalname).toLowerCase();
    if (!ALLOWED_IMAGE_EXTENSIONS.includes(extension)) {
      return {
        valid: false,
        error: 'File type not allowed. Allowed types: jpg, jpeg, png, gif, bmp, tiff, webp, ico, svg, heic, heif',
      };
    }

    if (!matchesMagicNumber(file.buffer, IMAGE_MAGIC_NUMBERS)) {
      return { valid: false, error: 'File content does not match image format' };
    }

    return { valid: true };
  }

  validateVideoFile(file: UploadedFileLike): { valid: boolean; error?: string } {
    if (!file || file.size === 0) return { valid: false, error: 'File is empty' };
    if (file.size > MAX_VIDEO_SIZE_BYTES) {
      return { valid: false, error: `File size exceeds ${MAX_VIDEO_SIZE_BYTES / (1024 * 1024)}MB limit` };
    }

    const extension = path.extname(file.originalname).toLowerCase();
    if (!ALLOWED_VIDEO_EXTENSIONS.includes(extension)) {
      return { valid: false, error: 'File type not allowed. Allowed types: mp4, webm, mov, avi, mkv, flv, wmv' };
    }

    if (!matchesMagicNumber(file.buffer, VIDEO_MAGIC_NUMBERS)) {
      return { valid: false, error: 'File content does not match video format' };
    }

    return { valid: true };
  }

  private async preparePath(
    originalFileName: string,
    kind: 'photos' | 'videos',
  ): Promise<{ fileName: string; dir: string; yearMonth: string }> {
    const fileName = generateFileName(originalFileName);
    const now = new Date();
    const yearMonth = `${now.getUTCFullYear()}/${String(now.getUTCMonth() + 1).padStart(2, '0')}`;
    const dir = path.join(this.basePath, kind, yearMonth);
    await fs.mkdir(dir, { recursive: true });
    return { fileName, dir, yearMonth };
  }
}

function generateFileName(originalFileName: string): string {
  return `${randomUUID().replace(/-/g, '')}${path.extname(originalFileName)}`;
}

function matchesMagicNumber(buffer: Buffer, allowed: number[][]): boolean {
  const header = buffer.subarray(0, 8);
  return allowed.some((magic) => magic.every((byte, i) => header[i] === byte));
}
